### ISSUE TITLE
Implement std::error::Error for parity_wasm::Error

### ISSUE THREAD

**Issue by fitzgen** [paritytech/parity-wasm#125]
This would allow for better integration with the larger ecosystem, for example:

```
error[E0277]: the trait bound `parity_wasm::Error: std::error::Error` is not satisfied
  --> src/lib.rs:33:18
   |
33 |     let module = Module::deserialize(reader)?;
   |                  ^^^^^^^^^^^^^^^^^^^^^^^^^^^^ the trait `std::error::Error` is not implemented for `parity_wasm::Error`
   |
   = note: required because of the requirements on the impl of `failure::Fail` for `parity_wasm::Error`
   = note: required because of the requirements on the impl of `std::convert::From<parity_wasm::Error>` for `failure::Error`
   = note: required by `std::convert::From::from`
```

**Comment by emk** [comment]
Yes, please! :-) This would make it much easier to use `parity-wasm` in an idiomatic fashion.

I see that there's a patch to implement this at 9fb26a9, but that doesn't seem to be in any of the obvious branches. Would anybody be interested in a PR implementing this feature?

**Comment by pepyakin** [comment]
i had some free time but unfortunately was interrupted. 

I would be very interested in a such PR!

**Comment by emk** [comment]
@pepyakin Thank you for the encouragement! Using your code as a starting point, I implemented `Error` for all the error types in #133. Please let me know if you'd like any further tweaks to the PR.
